README.md
Use [`bitsetium`](https://github.com/zakarumych/bitsetium) instead

Contact me if you want this crate name.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1]

### Fixed

Fixed set operation for layered bit set.

## [0.1.0]

### Added

- Traits for bitsets: `BitSet` and `BitSetMut` which allows mutation.
- Implementation of the traits above for primitive types.
- Implementation of the traits for various indirection types.
- Implementation of the traits for an `Option` type wrapper.
- Implementation for a layered bitset using top layer bitset as a marker to quickly which elements in bottom array contain at least one bit. Layers can be stacked indefinitely.
- Bunch of type aliases for bitsets of various sizes.


src/option.rs
use crate::{BitSet, BitSetMut};

impl<T> BitSet for Option<T>
where
    T: BitSet,
{
    const UPPER_BOUND: u32 = T::UPPER_BOUND;

    fn get(&self, index: u32) -> bool {
        match self {
            None => false,
            Some(bits) => bits.get(index),
        }
    }

    fn find_set(&self, lower_bound: u32) -> Option<u32> {
        match self {
            None => None,
            Some(bits) => bits.find_set(lower_bound),
        }
    }
}

impl<T> BitSetMut for Option<T>
where
    T: BitSetMut + Default,
{
    fn set(&mut self, index: u32, bit: bool) {
        match (&mut *self, bit) {
            (None, true) => {
                let mut bits = T::default();
                bits.set(index, true);
                *self = Some(bits);
            }
            (None, false) => {}
            (Some(bits), false) => {
                bits.set(index, false);
                if bits.is_empty() {
                    *self = None;
                }
            }
            (Some(bits), true) => bits.set(index, true),
        }
    }
}


src/layered.rs
use crate::{BitSet, BitSetMut};

#[derive(Debug)]
pub struct Layered<T, B, const N: usize> {
    top: T,
    bottom: [B; N],
}

impl<T, B, const N: usize> Default for Layered<T, B, N>
where
    T: Default + BitSet,
    B: Default + BitSet,
{
    fn default() -> Self {
        Self::new()
    }
}

impl<T, B, const N: usize> Layered<T, B, N>
where
    T: BitSet,
    B: BitSet,
{
    fn new() -> Self
    where
        T: Default,
        B: Default,
    {
        assert_eq!(T::UPPER_BOUND, N as u32);
        assert_eq!(T::UPPER_BOUND as usize, N);

        use core::mem::MaybeUninit;
        let bottom = unsafe {
            let mut array: [MaybeUninit<B>; N] = MaybeUninit::uninit().assume_init();
            for item in &mut array {
                // Leaks all previously written elements on panic. It is safe though.
                core::ptr::write(item, MaybeUninit::new(B::default()));
            }
            (&array as *const _ as *const [B; N]).read()
        };

        Layered {
            top: T::default(),
            bottom,
        }
    }
}

impl<T, B, const N: usize> BitSet for Layered<T, B, N>
where
    T: BitSet,
    B: BitSet,
{
    const UPPER_BOUND: u32 = B::UPPER_BOUND * N as u32;

    fn get(&self, index: u32) -> bool {
        assert!(index < Self::UPPER_BOUND);

        let t = index / B::UPPER_BOUND;
        let b = index % B::UPPER_BOUND;

        self.bottom[t as usize].get(b)
    }

    fn find_set(&self, lower_bound: u32) -> Option<u32> {
        assert!(lower_bound < Self::UPPER_BOUND);

        let t = lower_bound / B::UPPER_BOUND;
        let b = lower_bound % B::UPPER_BOUND;

        if b == 0 {
            let t = self.top.find_set(t)?;
            let b = self.bottom[t as usize].find_set(0)?;
            Some(t * B::UPPER_BOUND + b)
        } else {
            if self.top.get(t) {
                if let Some(b) = self.bottom[t as usize].find_set(b) {
                    return Some(t * B::UPPER_BOUND + b);
                }
            }

            let t = self.top.find_set(t + 1)?;
            let b = self.bottom[t as usize].find_set(0)?;
            Some(t * B::UPPER_BOUND + b)
        }
    }
}

impl<T, B, const N: usize> BitSetMut for Layered<T, B, N>
where
    T: BitSetMut,
    B: BitSetMut,
{
    fn set(&mut self, index: u32, bit: bool) {
        assert!(index < Self::UPPER_BOUND);
        let t = index / B::UPPER_BOUND;
        let u = index % B::UPPER_BOUND;

        if bit {
            if !self.top.get(t) {
                self.top.set(t, true);
            }
            self.bottom[t as usize].set(u, true)
        } else {
            if self.top.get(t) {
                self.bottom[t as usize].set(u, false);
                if self.bottom[t as usize].is_empty() {
                    self.top.set(t, false);
                }
            }
        }
    }
}


src/indirect.rs
use super::{BitSet, BitSetMut};

#[cfg(feature = "alloc")]
use alloc::boxed::Box;

impl<T> BitSet for &'_ T
where
    T: BitSet,
{
    const UPPER_BOUND: u32 = T::UPPER_BOUND;

    fn get(&self, index: u32) -> bool {
        T::get(*self, index)
    }

    fn find_set(&self, lower_bound: u32) -> Option<u32> {
        T::find_set(*self, lower_bound)
    }
}

impl<T> BitSet for &'_ mut T
where
    T: BitSet,
{
    const UPPER_BOUND: u32 = T::UPPER_BOUND;

    fn get(&self, index: u32) -> bool {
        T::get(*self, index)
    }

    fn find_set(&self, lower_bound: u32) -> Option<u32> {
        T::find_set(*self, lower_bound)
    }
}

impl<T> BitSetMut for &'_ mut T
where
    T: BitSetMut,
{
    fn set(&mut self, index: u32, bit: bool) {
        T::set(*self, index, bit)
    }
}

#[cfg(feature = "alloc")]
impl<T> BitSet for Box<T>
where
    T: BitSet,
{
    const UPPER_BOUND: u32 = T::UPPER_BOUND;

    fn get(&self, index: u32) -> bool {
        T::get(&**self, index)
    }

    fn find_set(&self, lower_bound: u32) -> Option<u32> {
        T::find_set(&**self, lower_bound)
    }
}

#[cfg(feature = "alloc")]
impl<T> BitSetMut for Box<T>
where
    T: BitSetMut,
{
    fn set(&mut self, index: u32, bit: bool) {
        T::set(&mut **self, index, bit)
    }
}


src/primitive.rs
use super::{BitSet, BitSetMut};

macro_rules! impl_for_primitive {
    ($ty:ty : $size:literal) => {
        impl BitSet for $ty {
            const UPPER_BOUND: u32 = $size;

            #[inline]
            fn get(&self, index: u32) -> bool {
                assert!(index < $size);
                0 != *self & ((1 as $ty) << index)
            }

            #[inline]
            fn find_set(&self, lower_bound: u32) -> Option<u32> {
                assert!(lower_bound < $size);
                let masked = *self & (!0 as $ty).wrapping_shl(lower_bound);
                match masked.trailing_zeros() {
                    $size => None,
                    index => Some(index),
                }
            }
        }

        impl BitSetMut for $ty {
            #[inline]
            fn set(&mut self, index: u32, bit: bool) {
                assert!(index < $size);
                match bit {
                    true => *self |= (1 as $ty << index),
                    false => *self &= !(1 as $ty << index),
                }
            }
        }
    };

    // ($ty:ty : $size:literal, $($tail_ty:ty : $tail_size:literal),+ $(,)?) => {
    //     impl_for_primitive!($($tail_ty : $tail_size),+);
    // }

    ($ty:ty : $size:literal, $($tail:tt)+) => {
        impl_for_primitive!($ty : $size);
        impl_for_primitive!($($tail)+);
    }
}

impl_for_primitive!(u8 : 8, u16 : 16, u32 : 32, u64 : 64, u128 : 128);


src/lib.rs
#![cfg_attr(not(test), no_std)]
#![deprecated(note = "use `bitsetium` crate instead")]

#[cfg(feature = "alloc")]
extern crate alloc;

mod indirect;
mod layered;
mod option;
mod primitive;

pub use layered::Layered;

/// Common trait for all bitsets in `bitseto` crate.
pub trait BitSet {
    /// Upper bound for this bitset indices.
    /// Implementation is expected to panic if index equal to or greter than `UPPER_BOUND` is speicifed.
    const UPPER_BOUND: u32;

    /// Returns bit state at specified index.
    ///
    /// # Panics
    ///
    /// This function may panic if `index` is equal to or greter than `UPPER_BOUND`.
    fn get(&self, index: u32) -> bool;

    /// Returns index of first set bit at index equal to or greter than specified.
    ///
    /// # Panics
    ///
    /// This function may panic if `index` is equal to or greter than `UPPER_BOUND`.
    fn find_set(&self, lower_bound: u32) -> Option<u32>;

    /// Returns true if no bit is set.
    fn is_empty(&self) -> bool {
        self.find_set(0).is_none()
    }
}

/// Common trait for all mutable bitsets in `bitseto` crate.
pub trait BitSetMut: BitSet {
    /// Sets bit state at specified index.
    ///
    /// # Panics
    ///
    /// This function may panic if `index` is equal to or greter than `UPPER_BOUND`.
    fn set(&mut self, index: u32, bit: bool);
}

pub type BitSet8 = u8;
pub type BitSet16 = u16;
pub type BitSet32 = u32;
pub type BitSet64 = u64;
pub type BitSet128 = u128;
pub type BitSet256 = layered::Layered<u32, u8, 32>;
pub type BitSet512 = layered::Layered<u64, u8, 64>;
pub type BitSet1024 = layered::Layered<u64, u16, 64>;
pub type BitSet2048 = layered::Layered<u64, u32, 64>;
pub type BitSet4096 = layered::Layered<u64, u64, 64>;
pub type BitSet8192 = layered::Layered<u64, u128, 64>;
pub type BitSet16384 = layered::Layered<u128, u128, 128>;

#[cfg(feature = "alloc")]
pub type BitSet32768 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet512>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet65536 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet1024>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet131072 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet2048>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet262144 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet4096>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet524288 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet8192>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet1048576 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet16384>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet2097152 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet32768>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet4194304 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet65536>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet8388608 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet131072>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet16777216 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet262144>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet33554432 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet524288>>, 64>;

#[cfg(feature = "alloc")]
pub type BitSet67108864 = layered::Layered<u64, Option<alloc::boxed::Box<BitSet1048576>>, 64>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn check_bounds() {
        assert_eq!(BitSet8::UPPER_BOUND, 8);
        assert_eq!(BitSet16::UPPER_BOUND, 16);
        assert_eq!(BitSet32::UPPER_BOUND, 32);
        assert_eq!(BitSet64::UPPER_BOUND, 64);
        assert_eq!(BitSet128::UPPER_BOUND, 128);
        assert_eq!(BitSet256::UPPER_BOUND, 256);
        assert_eq!(BitSet512::UPPER_BOUND, 512);
        assert_eq!(BitSet1024::UPPER_BOUND, 1024);
        assert_eq!(BitSet2048::UPPER_BOUND, 2048);
        assert_eq!(BitSet4096::UPPER_BOUND, 4096);
        assert_eq!(BitSet8192::UPPER_BOUND, 8192);
        assert_eq!(BitSet16384::UPPER_BOUND, 16384);
    }

    #[cfg(feature = "alloc")]
    #[test]
    fn check_large_bounds() {
        assert_eq!(BitSet32768::UPPER_BOUND, 32768);
        assert_eq!(BitSet65536::UPPER_BOUND, 65536);
        assert_eq!(BitSet131072::UPPER_BOUND, 131072);
        assert_eq!(BitSet262144::UPPER_BOUND, 262144);
        assert_eq!(BitSet524288::UPPER_BOUND, 524288);
        assert_eq!(BitSet1048576::UPPER_BOUND, 1048576);
        assert_eq!(BitSet2097152::UPPER_BOUND, 2097152);
        assert_eq!(BitSet4194304::UPPER_BOUND, 4194304);
        assert_eq!(BitSet8388608::UPPER_BOUND, 8388608);
        assert_eq!(BitSet16777216::UPPER_BOUND, 16777216);
        assert_eq!(BitSet33554432::UPPER_BOUND, 33554432);
        assert_eq!(BitSet67108864::UPPER_BOUND, 67108864);
    }

    #[cfg(feature = "alloc")]
    #[test]
    fn check_set() {
        let mut bits = Layered::<u64, u64, 64>::default();

        assert_eq!(bits.get(421), false);
        bits.set(421, true);
        assert_eq!(bits.get(421), true);

        bits.set(213, true);

        assert_eq!(bits.find_set(0), Some(213));
        assert_eq!(bits.find_set(300), Some(421));
        assert_eq!(bits.find_set(421 + 1), None);

        bits.set(213, false);
        assert_eq!(bits.find_set(0), Some(421));
        assert_eq!(bits.find_set(421 + 1), None);

        bits.set(421, false);
        assert_eq!(bits.get(421), false);
    }
}
